core/common/src/internal/math.kt

```diff
@@ -179,3 +179,63 @@ internal fun multiplyAndAdd(d: Long, n: Long, r: Long): Long {
     }
     return safeAdd(safeMultiply(md, n), mr)
 }
+
+internal val POWERS_OF_TEN = intArrayOf(
+    1,
+    10,
+    100,
+    1000,
+    10000,
+    100000,
+    1000000,
+    10000000,
+    100000000,
+    1000000000
+)
+
+/**
+ * The fraction [fractionalPart]/10^[digits].
+ */
+internal class DecimalFraction(
+    /**
+     * The numerator of the fraction.
+     */
+    val fractionalPart: Int,
+    /**
+     * The number of digits after the decimal point.
+     */
+    val digits: Int
+): Comparable<DecimalFraction> {
+    init {
+        require(digits >= 0) { "Digits must be non-negative, but was $digits" }
+    }
+
+    /**
+     * The integral numerator of the fraction, but with [newDigits] digits after the decimal point.
+     */
+    fun fractionalPartWithNDigits(newDigits: Int): Int = when {
+        newDigits == digits -> fractionalPart
+        newDigits > digits -> fractionalPart * POWERS_OF_TEN[newDigits - digits]
+        else -> (fractionalPart / POWERS_OF_TEN[digits - newDigits - 1] + 5) / 10
+    }
+
+    override fun compareTo(other: DecimalFraction): Int =
+        maxOf(digits, other.digits).let { maxPrecision ->
+            fractionalPartWithNDigits(maxPrecision).compareTo(other.fractionalPartWithNDigits(maxPrecision))
+        }
+
+    override fun equals(other: Any?): Boolean = other is DecimalFraction && compareTo(other) == 0
+
+    override fun toString(): String = buildString {
+        val denominator = POWERS_OF_TEN[digits]
+        append(fractionalPart / denominator)
+        append('.')
+        append((denominator + (fractionalPart % denominator)).toString().removePrefix("1"))
+    }
+
+    override fun hashCode(): Int {
+        var result = fractionalPart
+        result = 31 * result + digits
```

Comment:
Is it consistent with equals though?

Reply:
Thanks. It's not, and I think it would be a waste of time to try to come up with the correct one, so now an exception is thrown from `hashCode`.